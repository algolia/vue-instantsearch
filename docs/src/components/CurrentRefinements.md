---
title: CurrentRefinements
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/current-refinements.md
---

Show the currently refined values and allow them to be unset.

<a class="btn btn-static-theme" href="stories/?selectedKind=CurrentRefinements">ðŸ•¹ try out live</a>

## Usage

```html
<ais-current-refinements :clears-query="true"></ais-current-refinements>
```

## Props

Name | Type | Default | Description | Required
---|---|---|---|---
transformItems | Function | | Allows you to format and change the attributes | -
clearsQuery | Boolean | `false` | Should the 'clear all' button also clear the query? | -
excludedAttributes | Array | `[]` | Attributes not to show or clear | -
transformItems | `(items: object[]) => object[]` | `x => x` | Function which receives the items, which will be called before displaying them. Should return a new array with the same shape as the original array. Useful for mapping over the items to transform, remove or reorder them | -
classNames | Object | | Override class names | no

## Slots

name | scope | Description
---|---|---
`default` | `{ items: Item[], refine: Item => void, clearAll: () => void }` | Override how all the items look
`item`Â | `{ item: Item, refine: Item => void }` | Override how an item looks
`clearAllLabel`Â | `{ items: Item[] }` | Override how the "clear all" button looks

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Note that you can pass the prop `class-names`, with an object of class names and their replacement to override this.

Class name | Description
---|---
`ais-CurrentRefinements` | Container class
`ais-CurrentRefinements--noRefinement` | Modifier on the container when no refinements are enabled.
`ais-CurrentRefinements-item` | An item
`ais-CurrentRefinements-delete` | The button in the item to delete that refinement
`ais-CurrentRefinements-item--selected` | Selected item
