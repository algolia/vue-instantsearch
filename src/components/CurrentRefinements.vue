<template>
  <div :class="suit('')" v-if="state">
    <slot v-bind="state">
      <ul :class="suit('list')">
        <li v-for="item in state.refinements" :key="item.attributeName">
          <span :class="suit('item')">
            <span :class="suit('label')">{{item.attributeName | capitalize}}: {{item.computedLabel}}</span>
            <button :class="suit('delete')" @click="state.refine(item)">âœ•</button>
          </span>
        </li>
      </ul>
      <button :class="suit('reset')" @click="state.clearAllClick()" v-if="state.refinements.length > 0">
        Clear all
      </button>
    </slot>
  </div>
</template>

<script>
import algoliaComponent from '../component';
import { connectCurrentRefinedValues } from 'instantsearch.js/es/connectors';

export default {
  mixins: [algoliaComponent],
  props: {
    attributes: {
      type: Array,
      required: false,
    },
    clearsQuery: {
      type: Boolean,
      default: false,
      required: false,
    },
    excludedAttributes: {
      type: Array,
      required: false,
      validator(val) {
        // TODO: make this work with the connector (IS.js doesn't expose it)
        if (Array.isArray(val) && val.length > 0) {
          console.warn(
            '`excludedAttributes` is not implemented on CurrentRefinements'
          );
          return false;
        }
        return true;
      },
    },
  },
  data() {
    return {
      widgetName: 'CurrentRefinements',
    };
  },
  beforeCreate() {
    this.connector = connectCurrentRefinedValues;
  },
  computed: {
    widgetParams() {
      return {
        attributes: this.attributes,
        clearsQuery: this.clearsQuery,
      };
    },
  },
  filters: {
    capitalize(value) {
      if (!value) return '';
      return (
        value
          .toString()
          .charAt(0)
          .toUpperCase() + value.toString().slice(1)
      );
    },
  },
};
</script>
