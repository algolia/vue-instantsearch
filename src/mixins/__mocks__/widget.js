let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
