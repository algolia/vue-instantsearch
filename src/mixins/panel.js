import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
