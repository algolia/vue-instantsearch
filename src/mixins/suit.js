import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
