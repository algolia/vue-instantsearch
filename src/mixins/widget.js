const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
