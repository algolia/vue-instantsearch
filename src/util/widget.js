export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
