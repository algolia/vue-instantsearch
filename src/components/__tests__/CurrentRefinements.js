import { mount } from '@vue/test-utils';
import CurrentRefinements from '../CurrentRefinements.vue';
import { __setState } from '../../mixins/widget';

jest.mock('../../mixins/widget');
jest.mock('../../mixins/panel');

it('renders correctly (empty)', () => {
  __setState({
    items: [],
  });
  const wrapper = mount(CurrentRefinements);
  expect(wrapper.html()).toMatchSnapshot();
});

describe.each([
  [
    'facet',
    [
      {
        attribute: 'customFacet',
        label: 'customFacet',
        refinements: [
          {
            attribute: 'customFacet',
            type: 'facet',
            value: 'val1',
            label: 'val1',
          },
        ],
      },
    ],
  ],
  [
    'facet exclude',
    [
      {
        attribute: 'customExcludeFacet',
        label: 'customExcludeFacet',
        refinements: [
          {
            attribute: 'customExcludeFacet',
            type: 'exclude',
            value: 'val1',
            label: 'val1',
            exclude: true,
          },
        ],
      },
    ],
  ],
  [
    'disjunctive facet',
    [
      {
        attribute: 'customDisjunctiveFacet',
        label: 'customDisjunctiveFacet',
        refinements: [
          {
            attribute: 'customDisjunctiveFacet',
            type: 'disjunctive',
            value: 'val1',
            label: 'val1',
          },
        ],
      },
    ],
  ],
  [
    'hierarchical facet',
    [
      {
        attribute: 'customHierarchicalFacet',
        label: 'customHierarchicalFacet',
        refinements: [
          {
            attribute: 'customHierarchicalFacet',
            type: 'hierarchical',
            value: 'val1',
            label: 'val1',
          },
        ],
      },
    ],
  ],
  [
    'numeric filters',
    [
      {
        attribute: 'customNumericFilter',
        label: 'customNumericFilter',
        refinements: [
          {
            attribute: 'customNumericFilter',
            type: 'numeric',
            operator: '=',
            value: 'val1',
            label: '= val1',
          },
          {
            attribute: 'customNumericFilter',
            type: 'numeric',
            operator: '<=',
            value: 'val2',
            label: '≤ val2',
          },
          {
            attribute: 'customNumericFilter',
            type: 'numeric',
            operator: '>=',
            value: 'val3',
            label: '≥ val3',
          },
        ],
      },
    ],
  ],
  [
    'tag',
    [
      {
        attribute: '_tags',
        label: '_tags',
        refinements: [
          {
            attribute: '_tags',
            type: 'tag',
            value: 'tag1',
            label: 'tag1',
          },
        ],
      },
    ],
  ],
  [
    'query',
    [
      {
        attribute: 'query',
        label: 'query',
        refinements: [
          {
            attribute: 'query',
            type: 'query',
            value: 'search1',
            label: 'search1',
          },
        ],
      },
    ],
  ],
])('%s', (name, items) => {
  it('renders correct html', () => {
    __setState({
      items,
    });

    const wrapper = mount(CurrentRefinements);

    if (name === 'query') {
      expect(
        wrapper.find('.ais-CurrentRefinements-categoryLabel').contains('q')
      ).toBe(true);
    }

    expect(wrapper.html()).toMatchSnapshot();
  });
});

it('calls the Panel mixin with `canRefine`', () => {
  __setState({ items: [{}] });

  const wrapper = mount(CurrentRefinements);

  const mapStateToCanRefine = () =>
    wrapper.vm.mapStateToCanRefine(wrapper.vm.state);

  expect(mapStateToCanRefine()).toBe(true);

  wrapper.setData({
    state: {
      items: [],
    },
  });

  expect(mapStateToCanRefine()).toBe(false);

  expect(wrapper.vm.mapStateToCanRefine({})).toBe(false);
});

it('calls `refine` with a refinement', () => {
  const spies = [jest.fn(), jest.fn()];

  __setState({
    items: [
      {
        attribute: 'brands',
        label: 'brands',
        refine: spies[0],
        refinements: [
          {
            attribute: 'brands',
            type: 'facet',
            value: 'apple',
            label: 'apple',
          },
        ],
      },
      {
        attribute: 'colors',
        label: 'colors',
        refine: spies[1],
        refinements: [
          {
            attribute: 'colors',
            type: 'facet',
            value: 'green',
            label: 'green',
          },
        ],
      },
    ],
  });

  const wrapper = mount(CurrentRefinements);
  wrapper.find('.ais-CurrentRefinements-delete').trigger('click');

  expect(spies[0]).toHaveBeenLastCalledWith({
    attribute: 'brands',
    label: 'apple',
    type: 'facet',
    value: 'apple',
  });

  expect(spies[1]).not.toHaveBeenCalled();
});

describe('custom render', () => {
  const defaultScopedSlot = `
    <div slot-scope="{ refine, items, createURL }">
      <ul>
        <li
          v-for="item in items"
          :key="item.attribute"
          >
          {{item.label}}:
          <button
            v-for="refinement in item.refinements"
            @click="item.refine(refinement)"
            :key="refinement.value"
          >
            {{refinement.label}} ╳
          </button>
        </li>
      </ul>
    </div>
  `;

  const itemScopedSlot = `
    <div slot-scope="{ item, refine }">
      {{item.label}}:
      <button
        v-for="refinement in item.refinements"
        @click="item.refine(refinement)"
        :key="refinement.value"
      >
        {{refinement.label}}
      </button>
    </div>
  `;

  const refinementScopedSlot = `
    <div slot-scope="{ refinement, refine, createURL }">
      <button
        @click="refine(refinement)"
      >
        <pre>{{refinement}}</pre>
      </button>
    </div>
  `;

  const items = [
    {
      attribute: 'brands',
      label: 'brands',
      refinements: [
        {
          attribute: 'brands',
          type: 'facet',
          value: 'apple',
          label: 'apple',
        },
      ],
    },
    {
      attribute: 'colors',
      label: 'colors',
      refinements: [
        {
          attribute: 'colors',
          type: 'facet',
          value: 'red',
          label: 'red',
        },
      ],
    },
  ];

  it('gives all relevant info to scoped slot', () => {
    __setState({
      items,
      createURL: jest.fn(
        ({ attributeName, computedLabel }) =>
          `?${attributeName}=${computedLabel}`
      ),
    });

    const wrapper = mount(CurrentRefinements, {
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('has same amount of items', () => {
    __setState({
      items,
      createURL: jest.fn(
        ({ attributeName, computedLabel }) =>
          `?${attributeName}=${computedLabel}`
      ),
    });

    const wrapper = mount(CurrentRefinements, {
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    expect(wrapper.findAll('button')).toHaveLength(items.length);
  });

  it('item slot gives all relevant info to scoped slot', () => {
    __setState({
      items,
      createURL: jest.fn(
        ({ attributeName, computedLabel }) =>
          `?${attributeName}=${computedLabel}`
      ),
    });

    const wrapper = mount(CurrentRefinements, {
      scopedSlots: {
        item: itemScopedSlot,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('item slot has same amount of items', () => {
    __setState({
      items,
      createURL: jest.fn(
        ({ attributeName, computedLabel }) =>
          `?${attributeName}=${computedLabel}`
      ),
    });

    const wrapper = mount(CurrentRefinements, {
      scopedSlots: {
        item: itemScopedSlot,
      },
    });

    expect(wrapper.findAll('button')).toHaveLength(items.length);
  });

  it('refinement slot has same amount of items', () => {
    __setState({
      items,
      createURL: jest.fn(
        ({ attributeName, computedLabel }) =>
          `?${attributeName}=${computedLabel}`
      ),
    });

    const wrapper = mount(CurrentRefinements, {
      scopedSlots: {
        refinement: refinementScopedSlot,
      },
    });

    expect(wrapper.findAll('button')).toHaveLength(items.length);
  });
});
