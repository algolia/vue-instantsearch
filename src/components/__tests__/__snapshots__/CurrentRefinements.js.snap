// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom render gives all relevant info to scoped slot 1`] = `

<div class="ais-CurrentRefinements">
  <div>
    <ul>
      <li>
        brands:
        <button>
          apple ╳
        </button>
      </li>
      <li>
        colors:
        <button>
          red ╳
        </button>
      </li>
    </ul>
  </div>
</div>

`;

exports[`custom render item slot gives all relevant info to scoped slot 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <div>
        brands:
        <button>
          apple
        </button>
      </div>
    </li>
    <li class="ais-CurrentRefinements-item">
      <div>
        colors:
        <button>
          red
        </button>
      </div>
    </li>
  </ul>
</div>

`;

exports[`disjunctive facet renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        CustomDisjunctiveFacet:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          val1
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;

exports[`facet exclude renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        CustomExcludeFacet:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          val1
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;

exports[`facet renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        CustomFacet:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          val1
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;

exports[`hierarchical facet renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        CustomHierarchicalFacet:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          val1
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;

exports[`numeric filters renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        CustomNumericFilter:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          = val1
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          ≤ val2
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          ≥ val3
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;

exports[`query renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        Query:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          <q>
            search1
          </q>
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;

exports[`renders correctly (empty) 1`] = `

<div class="ais-CurrentRefinements ais-CurrentRefinements--noRefinement">
  <ul class="ais-CurrentRefinements-list">
  </ul>
</div>

`;

exports[`tag renders correct html 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li class="ais-CurrentRefinements-item">
      <span class="ais-CurrentRefinements-label">
        _tags:
      </span>
      <span class="ais-CurrentRefinements-category">
        <span class="ais-CurrentRefinements-categoryLabel">
          tag1
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
</div>

`;
